### DISCUSSION TITLE
Kustomization reconciling forever

### DISCUSSION THREAD

**Discussion by nicogl** [fluxcd/flux2#1843] [Q&A] [answered]
At some point in my CICD process I arrived to a state in which 2 of my Flux Kustomizations are ok and up to date, but one of them isn't.
For example, the `flux get ks` command shows below output in which _apps-kust_ is pointing to an old commit (95fd814fc0e33b533) instead of the latest one (de53d8db57a77bb74).

```
NAME            READY   MESSAGE                                                                REVISION                      SUSPENDED 
apps-kust       Unknown reconciliation in progress                                             master/95fd814fc0e33b533      False     
flux-system     True    Applied revision: master/decf3eefc39bdbe8967833ade53d8db57a77bb74      master/de53d8db57a77bb74      False    
infrastructure  True    Applied revision: master/decf3eefc39bdbe8967833ade53d8db57a77bb74      master/de53d8db57a77bb74      False
```

Every now and then I get to cases like this where Flux gets stuck and never recovers from that state...
I've tried reconciling the _apps_kust_ manually, even the entire _flux-system_ but it doesn't help.
The only way is to delete the Kustomization manually, which also removes its dependent HelmReleases causing outage, and creating it again. And what is worse... usually the kustomization deletion doesn't work (I guess because it's in "_reconciliation in progress_" state), so I end up removing components manually (pods, namespaces, etc) which is overkill and prone to error.

Is there anything I'm missing here?
The action that triggered all this was pushing a change to the CICD repository that updated a Helm Chart, which in turn triggered all the Flux deployment process...

**Comment by yebyen** [comment, marked as answer]
The default configuration of Kustomization is to set timeout at 5m0s and interval also at 5m0s

When the timeout perfectly lines up with the interval, it can appear that Kustomization is reconciling forever without ever failing or giving up to retry. Check if you have set custom values for interval and timeout, and to resolve this try making them different.

Timeout should probably be less than interval IMHO.

**Reply by stefanprodan** [reply]
In [v1beta2](https://github.com/fluxcd/kustomize-controller/pull/426) the timeout default has been changed to `interval - 30s` to avoid this forever reconciliation impression. Also the status message will display when the controller starts the health check and which is the timeout value.

**Reply by nicogl** [reply]
Oh good to know...
By the way, I'm using v1beta1, not v1beta2...

```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
```

I guess I should upgrade to the latest.

One more related thing. Sometimes after trying manual reconciliation for a Kustomization I get:
`context deadline exceeded`

What does this mean?

**Reply by stefanprodan** [reply]
It means that the CLI timeout is lower than the Kustomization timeout and the CLI timeouts before the controller gets into a ready state. Unless you set the Kustomization `spec.timeout` to a lower value than the interval, the CLI will timeout if your workloads are in crash loop.
